crates/net/src/handler.rs

```diff
@@ -0,0 +1,124 @@
+#![allow(unused)]
+//! Block Handler
+
+use alloy::primitives::Address;
+use libp2p::gossipsub::{IdentTopic, Message, MessageAcceptance, TopicHash};
+use std::time::SystemTime;
+use tokio::sync::watch::{channel, Receiver, Sender};
+// use ssz_rs::{prelude::*, List, Vector, U256};
+
+use crate::types::ExecutionPayloadEnvelope;
+
+/// This trait defines the functionality required to process incoming messages
+/// and determine their acceptance within the network. Implementors of this trait
+/// can specify how messages are handled and which topics they are interested in.
+pub trait Handler: Send {
+    /// Manages validation and further processing of messages
+    fn handle(&self, msg: Message) -> MessageAcceptance;
+
+    /// Specifies which topics the handler is interested in
+    fn topics(&self) -> Vec<TopicHash>;
+}
+
+/// Responsible for managing blocks received via p2p gossip
+#[derive(Debug, Clone)]
+pub struct BlockHandler {
+    /// Chain ID of the L2 blockchain. Used to filter out gossip messages intended for other
+    /// blockchains.
+    chain_id: u64,
+    /// A channel sender to forward new blocks to other modules
+    block_sender: Sender<ExecutionPayloadEnvelope>,
+    /// A [Receiver] to monitor changes to the unsafe block signer.
+    unsafe_signer_recv: Receiver<Address>,
+    /// The libp2p topic for pre Canyon/Shangai blocks.
+    blocks_v1_topic: IdentTopic,
+    /// The libp2p topic for Canyon/Delta blocks.
+    blocks_v2_topic: IdentTopic,
+    /// The libp2p topic for Ecotone V3 blocks.
+    blocks_v3_topic: IdentTopic,
+}
+
+impl Handler for BlockHandler {
+    /// Checks validity of a block received via p2p gossip, and sends to the block update channel if
+    /// valid.
+    fn handle(&self, msg: Message) -> MessageAcceptance {
+        tracing::debug!("received block");
+
+        if msg.topic == self.blocks_v1_topic.hash() {
+            // decode_pre_ecotone_block_msg::<ExecutionPayloadV1SSZ>(msg.data)
+            unimplemented!()
+        } else if msg.topic == self.blocks_v2_topic.hash() {
+            // decode_pre_ecotone_block_msg::<ExecutionPayloadV2SSZ>(msg.data)
+            unimplemented!()
+        } else if msg.topic == self.blocks_v3_topic.hash() {
+            // decode_post_ecotone_block_msg(msg.data)
+            unimplemented!()
+        } else {
+            return MessageAcceptance::Reject;
+        };
+
+        // match decoded {
+        //     Ok(envelope) => {
+        //         if self.block_valid(&envelope) {
+        //             _ = self.block_sender.send(envelope.payload);
+        //             MessageAcceptance::Accept
+        //         } else {
+        //             tracing::warn!("invalid unsafe block");
+        //             MessageAcceptance::Reject
+        //         }
+        //     }
+        //     Err(err) => {
+        //         tracing::warn!("unsafe block decode failed: {}", err);
+        //         MessageAcceptance::Reject
+        //     }
+        // }
```

Comment:
Curious about where we should put ssz types. Is there a better way to do this as opposed to rolling our own in this crate? These types are op-hardfork specific. See: https://github.com/a16z/magi/blob/master/src/network/handlers/block_handler.rs#L206-L208